### PR TITLE
[PF-2678] Implement --read-only flag for terra resource mount

### PR DESCRIPTION
**Pull request by rogerwangcs**
Support the `--read-only` flag for `terra resource mount`. Default behavior is to mount controller buckets created by the calling user as read/write and all other buckets/references as read only. If the user specifies `--read-only`, that value will override the default behavior.

Examples:
1. file-preview-bucket is created by another user
2. file-review-bucket is created by me (the calling user)

```
❯ terra resource mount
One or more resources failed to unmount.
❯ touch ~/workspace/file-preview-bucket/test.txt
touch: /Users/rogerwangcs/workspace/file-preview-bucket/test.txt: Read-only file system
❯ touch ~/workspace/another-file-review-bucket/test.txt
```

```
❯ terra resource mount --read-only=true
One or more resources failed to unmount.
❯ touch ~/workspace/file-preview-bucket/test.txt
touch: /Users/rogerwangcs/workspace/file-preview-bucket/test.txt: Read-only file system
❯ touch ~/workspace/another-file-review-bucket/test.txt
touch: /Users/rogerwangcs/workspace/another-file-review-bucket/test.txt: Read-only file system
```

```
❯ terra resource mount --read-only=false
One or more resources failed to unmount.
❯ touch ~/workspace/another-file-review-bucket/test.txt
❯ touch ~/workspace/file-preview-bucket/test.txt
```

### CODE REVIEW

File: src/main/java/bio/terra/cli/businessobject/resource/GcsObject.java [hunk 1]
```diff
@@ -150,15 +149,16 @@ public boolean isDirectory() throws SystemException {
...
-      Stream<BlobCow> objectsStream = StreamSupport.stream(objects.spliterator(), false);
 
       // Check for the presence of a directory object.
       // If a directory was created with the gcp cloud console, it will return false for
       // isDirectory(). In that case, we will check if there are more than 1 objects in the list.
       // This will correctly handle user created directories that are non-empty. Empty manually
       // create directories will not be detected as a directory as it's ambiguous from the API.
-      boolean isDirectory = objectsStream.anyMatch(object -> object.getBlobInfo().isDirectory());
-      long numObjects = objectsStream.count();
+      boolean isDirectory =
```
Comment (rogerwangcs) [inline src/main/java/bio/terra/cli/businessobject/resource/GcsObject.java:158]: Fix objectStream being used twice..

**Review by zloery — APPROVED** [review]

File: src/main/java/bio/terra/cli/command/resource/Mount.java [hunk 2]
```diff
@@ -26,6 +27,16 @@ public class Mount extends BaseCommand {
...
   private Boolean disableCache;
 
+  /**
+   * For controlled resources created by the calling user, the mount is always read-write regardless
+   * of the value of this flag.
+   *
+   * <p>For other resources, this flag will determine whether the mount is read-only or read-write,
+   * and defaults to read-only.
+   */
+  @CommandLine.Option(names = "--read-only", description = "Mount with only read permissions.")
```
Comment (yuhuyoyo) [inline src/main/java/bio/terra/cli/command/resource/Mount.java:37]: the comment you have above is great but it's not user facing, can you put the gist of it in the description?

**Review by yuhuyoyo — APPROVED** [review]

File: src/test/java/unit/MountHandlerTest.java [hunk 3]
```diff
@@ -127,7 +127,7 @@ public Collection<DynamicTest> mountGcsBucketTests() {
                   // Create a mountHandler and run mount
                   MountController mountController = MountControllerFactory.getMountController();
                   BaseMountHandler mountHandler =
-                      mountController.getMountHandler(gcsBucket, mountPath, false);
```
Comment (yuhuyoyo) [inline src/test/java/unit/MountHandlerTest.java:130]: nit: add comment for false e.g. /*readOnly=*/ false

File: src/test/java/unit/MountControllerTest.java [hunk 4]
```diff
@@ -97,6 +127,42 @@ void testMountResources() {
...
+      verify(spyMountController)
+          .getMountHandler(eq(resource1), eq(mountPath1), anyBoolean(), eq(false));
+      verify(spyMountController)
+          .getMountHandler(eq(resource2), eq(mountPath2), anyBoolean(), eq(true));
+
+      // Validate that readOnly flag overrides default mount permissions
+
+      // read only
+      spyMountController = getSpyMountController();
+      spyMountController.mountResources(false, true);
```
Comment (yuhuyoyo) [inline src/test/java/unit/MountControllerTest.java:150]: same here, not very readable what the false true are.

File: src/test/java/unit/MountControllerTest.java [hunk 5]
```diff
@@ -97,6 +127,42 @@ void testMountResources() {
...
+      MountController spyMountController = getSpyMountController();
+      spyMountController.mountResources(false, null);
+      verify(spyMountController)
+          .getMountHandler(eq(resource1), eq(mountPath1), anyBoolean(), eq(false));
+      verify(spyMountController)
+          .getMountHandler(eq(resource2), eq(mountPath2), anyBoolean(), eq(true));
+
+      // Validate that readOnly flag overrides default mount permissions
+
+      // read only
```
Comment (yuhuyoyo) [inline src/test/java/unit/MountControllerTest.java:148]: it's a real unit test, so let's split this into two.

mountResources_readOnly()
mountResources_readWrite()

File: src/main/java/bio/terra/cli/utils/mount/handlers/GcsFuseMountHandler.java [hunk 6]
```diff
@@ -12,31 +12,36 @@
 /** This class handles mounting a GCS bucket or prefix object using the GCS FUSE driver. */
 public class GcsFuseMountHandler extends BaseMountHandler {
 
-  private static final String FUSE_MOUNT_COMMAND = "gcsfuse";
+  private static final List<String> FUSE_MOUNT_COMMAND = List.of("gcsfuse", "--implicit-dirs");
```
Comment (yuhuyoyo) [inline src/main/java/bio/terra/cli/utils/mount/handlers/GcsFuseMountHandler.java:15]: add a comment about what is implicit-dirs

**Review by yuhuyoyo — APPROVED** [review]
some nits

File: src/main/java/bio/terra/cli/command/resource/Mount.java [hunk 7]
```diff
@@ -26,6 +27,16 @@ public class Mount extends BaseCommand {
...
+  private @Nullable Boolean readOnly;
```
[the remark this answers is not in the document]
Comment (rogerwangcs) [reply src/main/java/bio/terra/cli/command/resource/Mount.java:38]: Sorry I need to update the comments explaining this command as well, but there are 3 possible behaviors from this flag.
1. read-only is null - mount controlled resources created by calling user as read/write, others as read
2. read-only true - mount ALL resources as read
3. read-only false - mount ALL resources as read/write

File: src/test/java/unit/MountControllerTest.java (hunk 4 above)
Comment (rogerwangcs) [reply src/test/java/unit/MountControllerTest.java:150]: I forget that it's only readable in intellij since it auto annotates these params.

**Pull request merged**